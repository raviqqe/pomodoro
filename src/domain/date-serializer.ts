import { DateTime } from "luxon";

const DATE_FORMAT: string = "yyyyMMdd";

export class DateSerializer {
  public static serialize(date: Date): string {
    return DateTime.fromJSDate(date).toLocal().toFormat(DATE_FORMAT);
  }

  public static deserialize(formattedString: string): Date {
    return DateTime.fromFormat(formattedString, DATE_FORMAT).toJSDate();
  }
}
