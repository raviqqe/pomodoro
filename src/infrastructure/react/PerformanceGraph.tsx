import { last } from "lodash";
import { DateTime } from "luxon";
import {
  CartesianGrid,
  ResponsiveContainer,
  XAxis,
  YAxis,
  Label,
  Bar,
  BarChart,
} from "recharts";
import styled from "styled-components";
import { IPerformanceGraph } from "../../application/performance-graph";
import { DateSerializer } from "../../domain/date-serializer";
import { white } from "./style/colors";

const Container = styled.div`
  width: 80vw;
  height: 80vh;
`;

const Message = styled.div`
  color: ${white};
  font-size: 1.5em;
`;

export interface IProps {
  performanceGraph: IPerformanceGraph;
}

export const PerformanceGraph = ({
  performanceGraph: { data },
}: IProps): JSX.Element =>
  data.length === 0 ? (
    <Message>No performance graph to show yet!</Message>
  ) : (
    <Container>
      <ResponsiveContainer>
        <BarChart data={data}>
          <CartesianGrid fill="white" stroke="grey" strokeDasharray="3 3" />
          <XAxis
            dataKey="date"
            stroke="grey"
            tickFormatter={(date: string): string => {
              const days: number = DateTime.fromJSDate(
                DateSerializer.deserialize(
                  (last(data) as { date: string }).date
                )
              ).diff(
                DateTime.fromJSDate(DateSerializer.deserialize(date)),
                "days"
              ).days;

              return days === 0 ? "Today" : `${days} days ago`;
            }}
            tickMargin={10}
          />
          <YAxis allowDecimals={false} stroke="grey" tickMargin={5}>
            <Label
              angle={-90}
              position="insideLeft"
              style={{ fill: "grey" }}
              value="Pomodoros"
            />
          </YAxis>
          <Bar dataKey="pomodoros" fill="salmon" />
        </BarChart>
      </ResponsiveContainer>
    </Container>
  );
