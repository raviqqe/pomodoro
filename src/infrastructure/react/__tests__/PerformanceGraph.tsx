import { create } from "react-test-renderer";
import React from "react";
import { PerformanceGraph } from "../PerformanceGraph";
import { DateSerializer } from "../../../domain/date-serializer";

it("renders with no data", () => {
  expect(
    create(<PerformanceGraph performanceGraph={{ data: [] }} />).toJSON()
  ).toMatchSnapshot();
});

it("renders with data", () => {
  expect(
    create(
      <PerformanceGraph
        performanceGraph={{
          data: [
            { date: DateSerializer.serialize(new Date(0)), pomodoros: 42 },
          ],
        }}
      />,
      { createNodeMock: () => ({ parentElement: document.body }) }
    ).toJSON()
  ).toMatchSnapshot();
});
