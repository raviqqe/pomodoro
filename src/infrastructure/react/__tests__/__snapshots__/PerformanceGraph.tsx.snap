// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with data 1`] = `
<div
  className="sc-bdvvaa iodvJQ"
>
  <div
    className="recharts-responsive-container"
    style={
      Object {
        "height": "100%",
        "maxHeight": undefined,
        "minHeight": undefined,
        "minWidth": undefined,
        "width": "100%",
      }
    }
  />
</div>
`;

exports[`renders with no data 1`] = `
<div
  className="sc-gsDJrp bAPGAM"
>
  No performance graph to show yet!
</div>
`;
